rollout/trafficrouting/nginx/nginx.go

```diff
@@ -0,0 +1,216 @@
+	types "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/record"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/utils/defaults"
+	"github.com/argoproj/argo-rollouts/utils/diff"
+	logutil "github.com/argoproj/argo-rollouts/utils/log"
+)
+
+// Type holds this controller type
+const Type = "Nginx"
+
+// ReconcilerConfig describes static configuration data for the nginx reconciler
+type ReconcilerConfig struct {
+	Rollout        *v1alpha1.Rollout
+	Client         kubernetes.Interface
+	Recorder       record.EventRecorder
+	ControllerKind schema.GroupVersionKind
+}
+
+// Reconciler holds required fields to reconcile Nginx resources
+type Reconciler struct {
+	cfg ReconcilerConfig
+	log *logrus.Entry
+}
+
+// NewReconciler returns a reconciler struct that brings the canary Ingress into the desired state
+func NewReconciler(cfg ReconcilerConfig) *Reconciler {
+	return &Reconciler{
+		cfg: cfg,
+		log: logutil.WithRollout(cfg.Rollout),
+	}
+}
+
+// Type indicates this reconciler is an Nginx reconciler
+func (r *Reconciler) Type() string {
+	return Type
+}
+
+// canaryIngress returns the desired state of the canary ingress
+func (r *Reconciler) canaryIngress(stableIngress *extensionsv1beta1.Ingress, desiredWeight int32) (*extensionsv1beta1.Ingress, error) {
+	stableIngressName := r.cfg.Rollout.Spec.Strategy.Canary.TrafficRouting.Nginx.StableIngress
+	canaryIngressName := fmt.Sprintf("%s-canary", stableIngressName)
+	stableServiceName := r.cfg.Rollout.Spec.Strategy.Canary.StableService
+	canaryServiceName := r.cfg.Rollout.Spec.Strategy.Canary.CanaryService
+	annotationPrefix := defaults.GetCanaryIngressAnnotationPrefixOrDefault(r.cfg.Rollout)
+
+	// Set up canary ingress resource, we do *not* have to duplicate `spec.tls` in a canary, only
+	// `spec.rules`
+	desiredCanaryIngress := &extensionsv1beta1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        canaryIngressName,
+			Annotations: map[string]string{},
+		},
+		Spec: extensionsv1beta1.IngressSpec{
+			Rules: make([]extensionsv1beta1.IngressRule, 0), // We have no way of knowing yet how many rules there will be
+		},
+	}
+
+	// Must preserve ingress.class on canary ingress, no other annotations matter
+	// See: https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#canary
+	if val, ok := stableIngress.Annotations["kubernetes.io/ingress.class"]; ok {
+		desiredCanaryIngress.Annotations["kubernetes.io/ingress.class"] = val
+	}
+
+	// Ensure canaryIngress is owned by this Rollout for cleanup
+	desiredCanaryIngress.SetOwnerReferences([]metav1.OwnerReference{*metav1.NewControllerRef(r.cfg.Rollout, r.cfg.ControllerKind)})
+
+	// Copy only the rules which reference the stableService from the stableIngress to the canaryIngress
+	// and change service backend to canaryService. Rules **not** referencing the stableIngress will be ignored.
```

Comment:
Correct me if I am misunderstanding, but don't we want to keep the other rules that do not reference the stable service? Does the controller send a portion of **all** traffic to the config in the canary ingress or to only the routes with changes? I'm concerned that traffic will be routed to the canary config, but there is no route for it.

Reply:
I have validated that canaries are on a per host rule level.

You can have you main Ingress like this:
```yaml
spec:
  rules:
    - host: podinfo.example.com
      http:
        paths:
          - backend:
              serviceName: podinfo
              servicePort: 80
    - host: podinfo2.example.com
      http:
        paths:
          - backend:
              serviceName: podinfo2
              servicePort: 80
```

And then two separate canaries like this:
```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    nginx.ingress.kubernetes.io/canary: "true"
    nginx.ingress.kubernetes.io/canary-weight: "10"
  name: podinfo-canary
spec:
  rules:
  - host: podinfo.example.com
    http:
      paths:
      - backend:
          serviceName: podinfo-canary
          servicePort: 80
```

```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    nginx.ingress.kubernetes.io/canary: "true"
    nginx.ingress.kubernetes.io/canary-weight: "30"
  name: podinfo2-canary
spec:
  rules:
  - host: podinfo2.example.com
    http:
      paths:
      - backend:
          serviceName: podinfo2-canary
          servicePort: 80
```

So for the first IngressRule, we'd sent 10% to its canary, for the second one we'd send 30% to its canary.

And setting the code up the way I did allows for this scenario. Because someone might have a different `Rollout` using the stable `stableIngress`, but operating on a different `IngressRule` (read: Essentially two rollouts sharing an Ingress but different routes).

Reply:
That's a great point that I didn't think about! Okay, I'm good with this approach then. Thanks for testing confirming this behavior.